designer/server/src/main/scala/pl/touk/nussknacker/ui/db/timeseries/QuestDbFEStatisticsRepository.scala

```diff
@@ -0,0 +1,194 @@
+
+import java.io.File
+import java.nio.charset.StandardCharsets
+import java.nio.file.Files
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Try, Using}
+
+// TODO list:
+// 1. Writing in a single thread and polish usage of the wal2table job.
+// 2. Configurable directory of db (for now it's tmp directory) and limiting db file space on disk.
+// 3. Is truncate ok on critical errors?
+// 4. Collecting statistics should be deactivable by configuration.
+// 5. API should have better types (missing domain layer for FE statistics names).
+// 6. Handling errors while creating the QuestDb.
+// 7. Write tests when db directory is deleted in runtime.
+// 8. Changing table definition and recreate
+private class QuestDbFEStatisticsRepository(private val cairoEngine: CairoEngine)
+  extends FEStatisticsRepository[Future]
+    with LazyLogging {
+  private val ctx = new SqlExecutionContextImpl(cairoEngine, 1).`with`(AllowAllSecurityContext.INSTANCE, null)
+  private val applyWal2TableJob = new ApplyWal2TableJob(cairoEngine, 1, 1)
+
+  private lazy val statsWalTableWriter = {
+    cairoEngine.getWalWriter(cairoEngine.getTableTokenIfExists("statistics"))
+  }
+
+  private lazy val recordCursorFactory = {
+    cairoEngine.select(QuestDbFEStatisticsRepository.selectQuery, ctx)
+  }
+
+  override def write(statistics: Map[String, Long])
+                    (implicit ec: ExecutionContext): Future[Unit] = withExceptionHandling(
+    () => {
+      statistics.foreach { entry =>
+        val row = statsWalTableWriter.newRow(Os.currentTimeMicros())
+        row.putStr(0, entry._1)
+        row.putLong(1, entry._2)
+        row.append()
+      }
+      statsWalTableWriter.commit()
+      // This job is assigned to the WorkerPool created in Server mode (standalone db)
+      // (io.questdb.ServerMain.setupWalApplyJob, io.questdb.mp.Worker.run),
+      // so it's basically assigned thread to run this job.
+      // We will deal with that manually in our custom task to the periodic data flush.
+      // This method will be run by the task.
+      // TODO change to LazyList in scala 2.13
+      Range.inclusive(1, 3).takeWhile(idx => applyWal2TableJob.run(1) && idx <= 3)
+    },
+    ()
+  )
+
+  override def read()(implicit ec: ExecutionContext): Future[Map[String, Long]] = withExceptionHandling(
+    () => Using.resource(recordCursorFactory.getCursor(ctx)) { recordCursor =>
+      val buffer = ArrayBuffer.empty[(String, Long)]
+      val record = recordCursor.getRecord
+      while (recordCursor.hasNext) {
+        val name  = record.getStrA(0).toString
+        val count = record.getLong(1)
+        buffer.append(name -> count)
+      }
+      buffer.toMap
+    },
+    Map.empty[String, Long]
+  )
+
+  private def close(): Unit = {
+    recordCursorFactory.close()
+    statsWalTableWriter.close()
+    applyWal2TableJob.close()
+    cairoEngine.close()
```

Comment:
the engine not created in this class so the closing should be at the outside too